pkg/chunk/purger/blocks_purger.go

```diff
@@ -0,0 +1,234 @@
+		return
+	}
+	tManager := cortex_tsdb.NewTombstoneManager(api.bucketClient, userID, api.cfgProvider, api.logger)
+	deleteRequests, err := tManager.GetAllTombstonesForUser(ctx)
+	if err != nil {
+		level.Error(util_log.Logger).Log("msg", "error getting delete requests from the block store", "err", err)
+		http.Error(w, err.Error(), http.StatusInternalServerError)
+		return
+	}
+
+	if err := json.NewEncoder(w).Encode(deleteRequests); err != nil {
+		level.Error(util_log.Logger).Log("msg", "error marshalling response", "err", err)
+		http.Error(w, fmt.Sprintf("Error marshalling response: %v", err), http.StatusInternalServerError)
+	}
+
+	w.WriteHeader(http.StatusOK)
+
+}
+
+func (api *BlocksPurgerAPI) CancelDeleteRequestHandler(w http.ResponseWriter, r *http.Request) {
+	ctx := r.Context()
+	userID, err := tenant.TenantID(ctx)
+	if err != nil {
+		http.Error(w, err.Error(), http.StatusUnauthorized)
+		return
+	}
+
+	params := r.URL.Query()
+	requestID := params.Get("request_id")
+
+	tManager := cortex_tsdb.NewTombstoneManager(api.bucketClient, userID, api.cfgProvider, api.logger)
+	deleteRequest, err := tManager.GetTombstoneByIDForUser(ctx, requestID)
+	if err != nil {
+		level.Error(util_log.Logger).Log("msg", "error getting delete request from the object store", "err", err)
+		http.Error(w, err.Error(), http.StatusInternalServerError)
+		return
+	}
+
+	if deleteRequest == nil {
+		http.Error(w, "could not find delete request with given id", http.StatusBadRequest)
+		return
+	}
+
+	if deleteRequest.State == cortex_tsdb.StateCancelled {
+		http.Error(w, "the request has already been previously deleted", http.StatusBadRequest)
+		return
+	}
+
+	if deleteRequest.State == cortex_tsdb.StateProcessed {
+		http.Error(w, "deletion of request which is already processed is not allowed", http.StatusBadRequest)
+		return
+	}
+
+	if time.Since(deleteRequest.GetCreateTime()) > api.deleteRequestCancelPeriod {
+		http.Error(w, fmt.Sprintf("Cancellation of request past the deadline of %s since its creation is not allowed", api.deleteRequestCancelPeriod.String()), http.StatusBadRequest)
+		return
+	}
+
+	// create file with the cancelled state
+	_, err = tManager.UpdateTombstoneState(ctx, deleteRequest, cortex_tsdb.StateCancelled)
+	if err != nil {
+		level.Error(util_log.Logger).Log("msg", "error cancelling the delete request", "err", err)
+		http.Error(w, err.Error(), http.StatusInternalServerError)
+		return
+	}
+
+	w.WriteHeader(http.StatusNoContent)
+}
+
+func getTombstoneHash(startTime int64, endTime int64, selectors []string) string {
+	// Any delete request with the same start, end time and same selectors should result in the same hash
+
+	hash := md5.New()
+
+	bufStart := make([]byte, 8)
+	binary.LittleEndian.PutUint64(bufStart, uint64(startTime))
+	bufEnd := make([]byte, 8)
+	binary.LittleEndian.PutUint64(bufEnd, uint64(startTime))
+
+	hash.Write(bufStart)
+	hash.Write(bufEnd)
+
+	sort.Strings(selectors)
```

Comment:
I've updated it so it parses the matchers and then create the hash based on the parsed matchers. So any logically equivalent deletion requests should now result in the same hash. The example you provided should be the same hash. Also split up matchers like `match[]=process_start_time_seconds&match[]={job= "prometheus"}` would result in the same hash as the example you provided (Since a deletion request is the "AND" of all matchers provided).